### PR TITLE
fix(gatsby): trailing slash redirect in commands/serve

### LINKED ISSUE #9207: Preventing 301 redirects on URLs with no trailing slashes (Netlify)
## Summary

URLs with no trailing slash on sites hosted by Netlify lead to an immediate 301 redirect to the page with a trailing slash.

``foo.com/bar`` --> ``foo.com/bar/``

This has a performance cost and implications for SEO.

_Is there a Netlify configuration that resolves these URLs without redirecting?_

## Relevant information

While this question is specific to Netlify, I did a quick review of other Gatsby sites featured in the Showcase and saw the same behaviour in many, but not all cases, for example:

[Hopper /company](https://www.hopper.com/company) - 301 redirect (Netlify)
[Impossible Foods /mission](https://impossiblefoods.com/mission) - 301 redirect (unknown)
[Cajun Bow Fishing /bows](https://cajunbowfishing.com/category/bows) - 301 redirect (Netllify)
[Braun /shavers-for-men](https://ca.braun.com/en-ca/male-grooming/shavers-for-men) - 200 no redirect  (unknown)

### Environment (if relevant)

Same behaviour in Gatsby v1 and v2.
I'm using [gatsby-plugin-remove-trailing-slashes](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-plugin-remove-trailing-slashes) and [gatsby-plugin-netlify](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-plugin-netlify).
Within the project all Links point to the non-trailing slash version.

### LINKED ISSUE #19543: 301 redirects adding trailing slashes (gatsby build+serve)
Hello

## Description

I have problem with 301 redirects adding trailing slashes (/path => /path/) ONLY when I build+serve my site (gatsby build && gatsby serve && open http://localhost:8000/hi => I get redirected to http://localhost:8000/hi/). Doesn't happen using gatsby develop.

### Steps to reproduce

I have : 
- no gatsby-browser.js file
- no gatsby-ssr.js file
- an empty gatsby-config.js file
- gatsby-node.js
```
exports.createPages = ({ actions: { createPage } }) =>
  createPage({
    path: '/hi',
    component: HiComponent,
  });
```
- HiComponent renders hi and that's it

So... minimum code possible, and still have this issue. Tried gatsby-plugin-remove-trailing-slashes but didn't do anything, as it is only for pages created from the pages folder.

### Expected result

No 301, no redirect

### Actual result

Redirection

### Environment

```
System:
    OS: macOS 10.14.6
    CPU: (4) x64 Intel(R) Core(TM) i7-7567U CPU @ 3.50GHz
    Shell: 5.3 - /bin/zsh
  Binaries:
    Node: 12.12.0 - /var/folders/mt/71kkzssd3fq24s3x3y6wpc4h0000gn/T/yarn--1573821730999-0.42525738248781186/node
    Yarn: 1.19.1 - /var/folders/mt/71kkzssd3fq24s3x3y6wpc4h0000gn/T/yarn--1573821730999-0.42525738248781186/yarn
    npm: 6.11.3 - /usr/local/bin/npm
  Languages:
    Python: 2.7.16 - /usr/local/bin/python
  Browsers:
    Chrome: 78.0.3904.97
    Firefox: 65.0
    Safari: 13.0.3
```

Thanks in advance!

### PR DESCRIPTION
**Pull request by cosmn**
## Description
This is, in my opinion, a major issue at the core level of how gatsby generates and serves static files, usually blamed on Netlify and it's redirect settings.

My solution is a hack, and should be treated as such, until you'll find a better way to serve static files, another module , or a custom implementation, because you have a custom way of generating files.
I'm a first time Gatsby user, therefor I cannot asses the side-effects this patch may introduce, despite all tests passed successfully on my local machine. Your input here is highly required.

This PR is a continuation of an [abandoned PR#19567](https://github.com/gatsbyjs/gatsby/pull/19567) and my proposed solution matches, more or less, one of the approaches presented there by @pieh. I am basically deceiving `serve-static` into getting a trailing slash url, thus avoiding its redirect behavior.

## Related Issues

Fixes #19543
Assuming it also Fixes #9207

### CODE REVIEW

**Comment by cosmn** [comment]
Ping @pieh and @blainekasten because of their involvement in the matter

File: packages/gatsby/src/commands/serve.js [hunk 1]
```diff
@@ -40,6 +40,24 @@ const readMatchPaths = async program => {
...
+/**
+ * Express middleware that adds url trailing slash in order to
+ * trick static middleware and thus preventing redirect
+ * This needs to be used before express static middleware
+ * @param {*} req
+ * @param {*} res
+ * @param {*} next
+ */
+const directoryRedirectPrevention = (req, res, next) => {
+  if (req.url == `/` || req.url.indexOf(`.`) || req.url.slice(-1) !== `/`)
```
Comment (pvdz) [inline packages/gatsby/src/commands/serve.js]: Just something that caught my eye;

- `req.url.indexOf(`.`)` will only be falsy if the dot was the first character, otherwise it's always truthy since `indexOf` will return `-1` if the needle was not found.

I suspect you want to use `.includes('.')` here, instead. Considering this is very likely not what you wanted, I wonder how thoroughly this was tested and perhaps whether tests could be added for this. (Semantically, I wonder what the point is of checking for `'.'` in the first place. Can you please explain the idea behind that heuristic?)

- `req.url.slice(-1) !== '/'` Why are you returning early if the last character is not a forward slash?

Isn't that exactly what you want to fix? This seems a bit inverted.

- We use strict comparisons (`===` instead of `==`)

- We use brackets for multi-line statements (so `if (...) { ... }` unless everything is on the same line)

So basically, I wonder what cases are actually fixed by this code. Could you post an example of what this fixed for you?

Will leave the actual review for others mentioned in the PR.

Comment (cosmn) [reply packages/gatsby/src/commands/serve.js]: @pvdz thank you for catching that, indeed I meant to use `.includes()` instead of `.indexOf()`.
At the same time, you missed the entire issue and the solution I presented, Please refer to Issue #19543 and PR #19567. I really don't know how to explain this further.
I will go on and create a more descriptive code, one that follows the conduct you mentioned, containing the `includes` fix as well.

**Comment by blainekasten** [comment]
So i've spent more time researching this. I'm more convinced now that Gatsby is actually doing the right thing. For reference read this article: https://searchfacts.com/url-trailing-slash/

In addition, did you know that we have plugins for adding and removing slashes? I wonder if those would solve your needs instead of changing the way the codebase works. gatsby-plugin-remove-trailing-slash

**Comment by cosmn** [comment]
@pvdz I've made another commit to this issue, hopefully more descriptive, taking your feedback into consideration. This is not far from @pieh's ["200 redirect" proposal](https://github.com/gatsbyjs/gatsby/pull/19567#discussion_r347604007).

@blainekasten I appreciate your input, although I was not debating if url trailing slash is right or wrong. The issue here is an unexpected behavior, neither Gatsby or any other framework should perform redirects unless specifically instructed so. The author of Issue #19543 did a perfect job in replicating the behavior, which you confirmed with the [following comment](https://github.com/gatsbyjs/gatsby/issues/19543#issuecomment-554519778)

I also confirm that plugins like `gatsby-plugin-remove-trailing-slash` do nothing to fix the issue.

**Comment by cosmn** [comment]
A less intrusive solution would be to expose the `express` instance via gatsby-config, like you do with [developMiddleware](https://www.gatsbyjs.org/docs/api-proxy/#advanced-proxying), this time for production.

File: packages/gatsby/src/commands/serve.js [hunk 2]
```diff
@@ -40,6 +41,29 @@ const readMatchPaths = async program => {
...
+ * This is an express middleware to be used before static middleware
+ * @param {*} req
+ * @param {*} res
+ * @param {*} next
+ */
+const directoryRedirectPrevention = (req, res, next) => {
+  // url is home page
+  if (req.url === `/`) return next()
+  // url has trailing slash, no need to intervene
+  if (req.url.slice(-1) === `/`) return next()
```
Comment (pvdz) [inline packages/gatsby/src/commands/serve.js]: The second check subsumes the first because `'x'.slice(-1) === 'x'` :)

File: packages/gatsby/src/commands/serve.js [hunk 3]
```diff
@@ -40,6 +41,29 @@ const readMatchPaths = async program => {
...
+  // url is a file request, no need for trailing slash
+  if (fileExtensionRegEx.test(req.url)) return next()
```
Comment (pvdz) [inline packages/gatsby/src/commands/serve.js]: This seems to overshoot a little bit. This is basically the `.includes()` check while trying to assert that there is a non-dot before and after the dot. I don't have enough context but is there a particular reason for this?

Also, I think the `$` in the regex serves no purpose? (I don't think there's a difference with and without it). If you're set on keeping it, I think `/.\../` is the same regex. Additionally, this assumes folder parts can not have a dot (that may be legit, depending on the source of urls, but is certainly not generic).

File: packages/gatsby/src/commands/serve.js [hunk 4]
```diff
@@ -40,6 +41,29 @@ const readMatchPaths = async program => {
...
+
+  // turning /slug into /slug/
+  req.originalUrl += `/` // https://expressjs.com/en/api.html#req.originalUrl
```
Comment (pvdz) [inline packages/gatsby/src/commands/serve.js]: The first example in this link has:
```
// GET /search?q=something
console.dir(req.originalUrl)
// => '/search?q=something'
```

So potentially this line would be doing `originalUrl = '/search?q=something' + '/'`. That's probably not what you wanted?

File: packages/gatsby/src/commands/serve.js [hunk 5]
```diff
@@ -40,6 +41,29 @@ const readMatchPaths = async program => {
...
+  req.url += `/`
```
Comment (pvdz) [inline packages/gatsby/src/commands/serve.js]: Which made me look into `.url`, so. Same link has

> req.url is not a native Express property, it is inherited from Node’s http module.

> https://nodejs.org/api/http.html#http_message_url
> > When request.url is '/status?name=ryan'

So it looks like `.url` has the same problem. Is there a reason why this is not a problem?

File: packages/gatsby/src/commands/serve.js (hunk 2 above)
Comment (cosmn) [reply packages/gatsby/src/commands/serve.js]: That's true, I wanted to differentiate the two cases, easier to understand.
Falls into "write code for humans" practice. The fact that you noticed, proves my point.

File: packages/gatsby/src/commands/serve.js (hunk 4 above)
Comment (cosmn) [reply packages/gatsby/src/commands/serve.js]: Indeed, url's with query args is a use case I haven't tackled yet. I'm not fully familiar with gatsby's handle of query args. Your input here is required.
Do we turn `/search?q=something` into `/search/?q=something` ?

File: packages/gatsby/src/commands/serve.js (hunk 2 above)
Comment (pvdz) [reply packages/gatsby/src/commands/serve.js]: I get what you're saying but I still think I'd prefer to drop the redundant check and updating the comment to reflect the two cases being caught here.

File: packages/gatsby/src/commands/serve.js (hunk 4 above)
Comment (pvdz) [reply packages/gatsby/src/commands/serve.js]: To be honest, I have no idea :)

Maybe @pieh knows

Comment (pieh) [reply packages/gatsby/src/commands/serve.js]: We have this piece of code in production runtime that adjust url to use canonical page path for initial page load (so this will be connected with to `gatsby serve` behaviour in some way) - https://github.com/gatsbyjs/gatsby/blob/9d3a8f080ca0a9d58a92a978931b3438f4d64e0d/packages/gatsby/cache-dir/production-app.js#L99-L119

Just note that "canonical path" on its own doesn't specify if it has trailing slash or not - it just mean `path` that was used with `createPage` action call (Gatsby doesn't normalize those right now - at least for pages created in user code in `createPages` hook). Gatsby does however use trailing slashes for pages created from `src/pages` directory (so it does have opinion here).

I'm not sure if this make any sense or if this is answer any question here.

But for manipulating URLs I recommend using builtin Node's `url` module - particularly:
 - [url.parse](https://nodejs.org/api/url.html#url_url_parse_urlstring_parsequerystring_slashesdenotehost) - to created structured url object
 - then modify `pathname` part to add trailing slash
 - [url.format](https://nodejs.org/api/url.html#url_url_format_urlobject) - for converting object back to string

**Comment by pvdz** [comment]
@cosmn Hi :) Are you still working on this? The PR hasn't seen an update in the past month so I hope you're still looking into this. If you're no longer interested, could you please let us know all the same? Thanks!

**Comment by cosmn** [comment]
Hi @pvdz. I apologize for the late reply; busy period
I stopped working on this once I realized its drifting away from a constructive discussion on the solution at hand. I suggest to revisit your policy of welcoming and handling PRs. It felt hostile and dismissive. You were nitpicking code syntax instead of cooperating.

Since I will not contributing with code anymore, but I still want to see this issue addressed and fixed, I will write here what I think is the root cause of the entire situation:

Gasby's approach of generating static files, conflicts with the way [serve-static](https://github.com/expressjs/serve-static) is handling directories.
1. Either programmatically through `createPage` or other internal means, Gatsby will generate a page as `/public/generated-page-slug/index.html`. Nothing wrong with that, except that:
2. `serve-static` performs a redirect when it encounters a directory, as seen in their [source code](https://github.com/expressjs/serve-static/blob/94feedb81682f4503ed9f8dc6d51a5c1b9bfa091/index.js#L182-L210)
```javascript
68  var onDirectory = redirect
69      ? createRedirectDirectoryListener()
70      : createNotFoundDirectoryListener()
...
99  stream.on('directory', onDirectory)
...
182 function createRedirectDirectoryListener () {
183     return function redirect (res) {
...
```
Although `serve-static` has part of the blame here for creating an unexpected behavior, they wont be able to fix this due to the high amount of dependents relying on it, projects and node modules.
The rest of the blame is on Gatsby for not having its own static server module, which can be controlled, improved, exposed to developers for extended functionality.
At least save those files to `/public/generated-page-slug.html`, most likely under a flag to avoid a breaking change.
I find it hard to believe this behavior hasn't arise during development and tests. It just seems swept under the carpet for way too long

People have created several issues and countless comments on this particular matter. Whoever is familiar enough with this part of Gatsby's functionality, Please make things right.

You may close this PR when you see fit.

**Comment by wardpeet** [comment]
Could you tell me

Hey @cosmn,

Sorry about the above, you have to understand that we have to also keep our code maintainable. Our apologies if it didn't feel welcoming at all.

I do have a question for you that I also added in other PR that does similar things.

See https://github.com/gatsbyjs/gatsby/pull/19567#issuecomment-620211002
> > So really gatsby is hiding the redirects the server is doing by doing its own clientside redirects. Our team had a similar issue with gatsby before, and the google SEO tools told us about it.

> I'm aware of gatsby is hiding this implementation and that a redirection is happening. The problem is that when we build, we always build like you have trailing slashes and it's the server implementation that needs to handle pretty URLs, not us.

> `gatsby serve` is not a production server, it's merely there to run e2e-tests, do a final check if everything is looking alright.

> So I'm still unsure what the use case is of not doing a 301 in serve.

**Comment by ascorbic** [comment]
I'm going to close this PR as it doesn't seem to be going anywhere
